arch: arm: cortex_m: maintainability improvement by reducing assembly surface

ithinuel opened the pull request:
This PR simplifies maintenance and improves portability by relying on C where asm is not strictly required.

This requires:

- #64978

ithinuel committed:
arch: arm: cortex_m: use C rather than asm in isr_wrapper & exc_exit

Asm is notoriously harder to maintain than C and requires core specific
adaptation which impairs even more the readability of the code.

This is a first step in reducing the amount of ASM in arch/arm/cortex_m

Signed-off-by: Wilfried Chauveau <wilfried.chauveau@arm.com>

ithinuel committed:
arch: arm: cortex_m: Use outlined arch_irq_(un)lock in assembly

Asm is notoriously harder to maintain than C and requires core specific
adaptation which impairs even more the readability of the code.

This change reduces the need for core specific conditional compilation.

Signed-off-by: Wilfried Chauveau <wilfried.chauveau@arm.com>

ithinuel committed:
arch: arm: cortex_m: Document why __aeabi_read_tp impl requires ASM impl

This method has special ABI requirement that requires the use of ASM.
This change documents why this is required & adds reference to the
related specification.

Signed-off-by: Wilfried Chauveau <wilfried.chauveau@arm.com>

ithinuel committed:
arch: arm: cortex_m: use cmsis API rather than assembly

Asm is notoriously harder to maintain than C and requires core specific
adaptation which impairs even more the readability of the code.

Signed-off-by: Wilfried Chauveau <wilfried.chauveau@arm.com>

ithinuel committed:
arch: arm: cortex_m: Use r* register names rather than v*

v* register aliases are uncommon and it can be surprising to find them.
This change makes use of r* register names for a more consistent
experience of reading assembly.

Signed-off-by: Wilfried Chauveau <wilfried.chauveau@arm.com>

ithinuel committed:
arch: arm: cortex_m: Use arch_irq_lock instead of a bespoke asm impl

This change is another step in unifying the use of irq_locks on arch arm
targets.

Signed-off-by: Wilfried Chauveau <wilfried.chauveau@arm.com>

ithinuel committed:
arch: arm: cortex_m: move part of swap_helper to C

Asm is notoriously harder to maintain than C and requires core specific
adaptation which impairs even more the readability of the code.

This change significantly enhances the maintainability & portability of the
code at the expanse of an indirection (1 outlined function).

Signed-off-by: Wilfried Chauveau <wilfried.chauveau@arm.com>

erwango commented on arch/arm/core/cortex_m/isr_wrapper.c:
```diff
@@ -0,0 +1,88 @@
+	__enable_irq();
+#endif /* CONFIG_PM */
+
+#if defined(CONFIG_ARM_CUSTOM_INTERRUPT_CONTROLLER)
```
This isn't related to your change but this option is defined as :
```
config ARM_CUSTOM_INTERRUPT_CONTROLLER
	bool
	depends on !CPU_CORTEX_M
```
so I wonder if this should really be kept, or if there is some subtlety that I don't catch.

ithinuel replied:
It might be a missed part on my end. There has been some work on that topic on upstream while I was working on this PR locally.

ithinuel replied:
It doesn't seem to me that the snippet you quoted is correct: https://github.com/zephyrproject-rtos/zephyr/blob/main/arch/arm/Kconfig#L17-L35

erwango replied:
ok, right. So "There has been some work on that topic on upstream while I was working on this PR locally." actually applies to my review. Sorry for the noise.

microbuilder commented:
@ithinuel Can you rebase this? Thanks.

ithinuel committed:
arch: arm: cortex_m: Use cmsis api instead of inline asm in arch_irq_*

Asm is notoriously harder to maintain than C and requires core specific
adaptation which impairs even more the readability of the code.

Signed-off-by: Wilfried Chauveau <wilfried.chauveau@arm.com>

ithinuel committed:
arch: arm: cortex_m: Convert cpu_idle from ASM to C

Asm is notoriously harder to maintain than C and requires core specific
adaptation which impairs even more the readability of the code.

This change reduces the need for core specific conditional compilation and
unifies irq locking code.

Signed-off-by: Wilfried Chauveau <wilfried.chauveau@arm.com>

# Conflicts:
#	soc/arm/nordic_nrf/nrf53/soc_cpu_idle.h

ithinuel commented:
@microbuilder I rebased this PR.

ithinuel committed:
arch: arm: cortex_m: Only trigger context switch if thread is preemptible

This is a fix for #61761 where a cooperative task is switched from at the
end of an exception. A cooperative thread should only be switched from if
the thread exists the ready state.

Signed-off-by: Wilfried Chauveau <wilfried.chauveau@arm.com>

ithinuel commented:
Builder 9 & 10 show failure on qemu_x86_64, riscv32 and riscv64.
They don’t seem related to the changes in this PR. I’ll rebase just in case it’s been already fixed upstream.

microbuilder commented:
@carlescufi There's some Nordic-specfiic changes in here someone from Nordic should probably look at.

Does it make sense to try to get these in now before the feature freeze, or wait until the start of the LTS dev cycle in your opinion?

ithinuel commented:
I'd like to have it as soon as possible to have as much time as possible to stabilise it before next LTS.

carlescufi commented:
@anangl  and @gmarull  can you please take a look?

microbuilder commented:
> I'd like to have it as soon as possible to have as much time as possible to stabilise it before next LTS (just in case there's some unexpected oversight).

I just want to make sure you're available during the release cycle, then, to make changes quickly if required, since this is a major change and could turn into a blocker for the release.

MaureenHelm commented:
@ithinuel please rebase

ithinuel commented:
@MaureenHelm (finally) rebased :)

carlescufi commented:
Merging since the maintainer is the author.

ithinuel commented:
Hello @Thalley, I am aware of some breakage on some NXP platforms (see [#64784](https://github.com/zephyrproject-rtos/zephyr/issues/64784#issuecomment-2058625219)) but I’m not aware of anything else.

Can you please create an issue & provide a method to reproduce?

gchwier commented:
It causes also a rolling reboot in tests of MCUboot.
To reproduce:
`west build --sysbuild -p -b nrf5340dk/nrf5340/cpuapp zephyr/tests/boot/with_mcumgr`
`west flash`

gchwier commented:
when added 
```
	while (1) {
		k_sleep(K_MSEC(1000));
	}
```
instead  of just returning main, it works (no rolling reboot)

carlescufi commented:
@gchwier can you please open an issue as well?

ithinuel commented:
@gchwier I think I identified an issue in `cpu_idle.c`.

Would you mind checking if https://github.com/zephyrproject-rtos/zephyr/pull/71593 solves it?
This is the only significant difference I could notice between the two generated binaries.

Thalley commented:
> @gchwier I think I identified an issue in `cpu_idle.c`.
> 
> Would you mind checking if #71593 solves it? This is the only significant difference I could notice between the two generated binaries.

It does not fix the issue I'm seeing - I'll create a bug report soon

Thalley commented:
> > @gchwier I think I identified an issue in `cpu_idle.c`.
> > Would you mind checking if #71593 solves it? This is the only significant difference I could notice between the two generated binaries.
> 
> It does not fix the issue I'm seeing - I'll create a bug report soon

@ithinuel https://github.com/zephyrproject-rtos/zephyr/issues/71596

Interestingly there are no issues for non-Bluetooth applications, so may be related to the HCI driver or something else.

nordic-piks commented:
> > > @gchwier I think I identified an issue in `cpu_idle.c`.
> > > Would you mind checking if #71593 solves it? This is the only significant difference I could notice between the two generated binaries.
> > 
> > 
> > It does not fix the issue I'm seeing - I'll create a bug report soon
> 
> @ithinuel #71596
> 
> Interestingly there are no issues for non-Bluetooth applications, so may be related to the HCI driver or something else.

There are issues for multiple tests (not only Bluetooth), eg.
nrf9160dk_nrf9160: tests/kernel/workq/work_queue/kernel.workqueue
`10:12:56  ERROR   - *** Booting Zephyr OS build v3.6.0-2572-gbec4faebb9d5 ***
10:12:56  Running TESTSUITE workqueue_delayed
10:12:56  ===================================================================
10:12:56  ASSERTION FAIL [!z_is_thread_state_set(_kernel.cpus[0].current, ((1UL << (4))))] @ ../../../../../../../zephyr/kernel/sched.c:1170
10:12:56  E: ***** HARD FAULT *****
10:12:56  E:   Fault escalation (see below)
10:12:56  E: ARCH_EXCEPT with reason 4
10:12:56  E: r0/a1:  0x00000004  r1/a2:  0x00000492  r2/a3:  0x00000001
10:12:56  E: r3/a4:  0x00000004 r12/ip:  0x00000000 r14/lr:  0x00006c09
10:12:56  E:  xpsr:  0x29000000
10:12:56  E: Faulting instruction address (r15/pc): 0x000086d0
10:12:56  E: >>> ZEPHYR FATAL ERROR 4: Kernel panic on CPU 0
10:12:56  E: Current thread: 0x200008f0 (main)
10:12:56  E: Halting system`

gchwier commented:
> @gchwier I think I identified an issue in `cpu_idle.c`.
> 
> Would you mind checking if #71593 solves it? This is the only significant difference I could notice between the two generated binaries.

With that fix there is not rolling-boot, but there is also no printed messages on the console, and shell is not available

stephanosio commented on arch/arm/core/cortex_m/swap_helper.S:
```diff
@@ -117,125 +118,20 @@ out_fp_endif:
-    orrs r3, r0
-    msr CONTROL, r3
-
-    /* ISB is not strictly necessary here (stack pointer is not being
-     * touched), but it's recommended to avoid executing pre-fetched
-     * instructions with the previous privilege.
-     */
-    isb
```
Important bits like these get lost when translating to C.

stephanosio reviewed (commented):
I am a bit late to the party here ... but is converting the core arch implementations written in ASM to C really something we want/should be doing?

The `_isr_wrapper` function in `isr_wrapper.S`, for instance, is called for every interrupt  and just a few instructions added to it can lead to a significant performance degradation. Same goes for `z_arm_pendsv`, which is called for every context switch.

Places like these are where we want complete control over the exact instructions used and hence why they were written in ASM.

Also, it is more difficult to control data and instruction barriers by switching to the C implementation (especially using the CMSIS functions), and this can lead to unnecessary and redundant pipeline flushes and hence performance degradation.

Moreover, the existing ASM implementation was fairly clean, well commented, and easy enough to read.

Given that this has already caused many issues for us and will likely cause more in the future, should we look into reverting it and getting back to the good old ASM implementation?

ithinuel commented:
Thank you for your valuable feedback.
The idea of this change was initiated by the difficulties encountered while learning about zephyr and its implementation.

I understand all your points and agree with most of them. However, since its introduction, the code has evolved to include a significant amount of conditional features inclusion which compounds with accommodations for the various architectural variants (v6m, v7m, v8m.main, v8m.base etc) makes some of the code difficult to read and keep track of.
[`isr_wrapper.S`](https://github.com/zephyrproject-rtos/zephyr/blob/a2ed8162c8c42a4301264709e9e476ef9eaa0552/arch/arm/core/cortex_m/isr_wrapper.S) literally had nearly as much instruction lines (36) as pre-processor (30), and this is counting the blocks like [this](https://github.com/zephyrproject-rtos/zephyr/blob/a2ed8162c8c42a4301264709e9e476ef9eaa0552/arch/arm/core/cortex_m/isr_wrapper.S#L78-L94) as 8instructions for 5pre-processor.

In the various comparison I made, the generated code was on par or better than the previous solution. The most recent comparison I did was built with `west build --sysbuild -p -b nrf5340dk/nrf5340/cpuapp tests/boot/with_mcumgr` at a2ed8162c8c and 4b276c49620 where the generated code for `_isr_wrapper` is effectively 1 instruction smaller than the ASM implementation.

In a number of places, the generated code is/was either suboptimal or incorrect. Eg in [this](https://github.com/zephyrproject-rtos/zephyr/blob/main/arch/arm/core/cortex_m/swap_helper.S#L190-L208) instance (which hasn’t been changed as part of this PR), it would fail to build for Cortex-M23 with `BUILTIN_STACK_GUARD` enabled and will generate sub-optimal code with unnecessary pop/push sequences. So the ASM implementation is not free from latent issues and I obviously intend to fix these aswell in upcoming contributions.

A notable pain point is around manual in-lining of irq locking/unlocking which makes it hard to keep them all consistent and to introduce changes. Eg, the implementation of SMP on Armv6-m will require to
enable SoC specific synchronisation between core. Identifying and propagating the required changes will be challenging.
The issues raised so far are actually all related to missing calls `__enable_irq()`/`__disable_irq()`.

Due to the nature of the change, I was expecting some breakage to have sipped through the checks (all CI passed). I am still ready to investigate and document the cause of any issue that may still arise. Hopefully, this hasn’t inconvenienced too many people and I appreciate the patience of those who have been.

So I don’t think reverting would be an appropriate move and I remain convinced this is a good thing for the maintainability.
